plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,373 @@
+                writeFuture.addListener(() -> bufferPool.offer(buffer), directExecutor());
+                addExceptionCallback(writeFuture, throwable -> failure.compareAndSet(null, throwable));
+            }
+        }
+    }
+
+    @ThreadSafe
+    private static class BufferPool
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(BufferPool.class).instanceSize();
+
+        private final int numBuffers;
+        private final long bufferRetainedSize;
+        @GuardedBy("this")
+        private final Queue<SliceOutput> freeBuffersQueue;
+        @GuardedBy("this")
+        private CompletableFuture<Void> blockedFuture = new CompletableFuture<>();
+        private volatile boolean closed;
+
+        public BufferPool(int numBuffers, int writeBufferSize)
+        {
+            this.numBuffers = numBuffers;
+            this.freeBuffersQueue = new ArrayDeque<>(numBuffers);
+            for (int i = 0; i < numBuffers; ++i) {
+                freeBuffersQueue.add(Slices.allocate(writeBufferSize).getOutput());
+            }
+            if (freeBuffersQueue.isEmpty()) {
+                this.bufferRetainedSize = 0;
+            }
+            else {
+                this.bufferRetainedSize = freeBuffersQueue.peek().getRetainedSize();
+            }
+        }
+
+        public synchronized CompletableFuture<Void> isBlocked()
+        {
+            if (freeBuffersQueue.isEmpty()) {
+                if (blockedFuture.isDone()) {
+                    blockedFuture = new CompletableFuture<>();
+                }
+                return blockedFuture;
+            }
+            else {
+                return NOT_BLOCKED;
+            }
+        }
+
+        public synchronized SliceOutput take()
+        {
+            while (true) {
+                if (closed) {
+                    return null;
+                }
+                if (!freeBuffersQueue.isEmpty()) {
+                    return freeBuffersQueue.poll();
+                }
+                try {
+                    wait();
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+
+        public void offer(SliceOutput buffer)
+        {
+            if (closed) {
+                return;
+            }
+
+            buffer.reset();
+
+            CompletableFuture<Void> completableFuture;
+            synchronized (this) {
+                completableFuture = blockedFuture;
+                freeBuffersQueue.add(buffer);
+                notify();
+            }
+
+            completableFuture.complete(null);
+        }
+
+        public long getRetainedSize()
+        {
+            if (closed) {
+                // no in-flight buffers at this point
+                synchronized (this) {
+                    return INSTANCE_SIZE + freeBuffersQueue.size() * bufferRetainedSize;
```

Comment:
we should expect `freeBuffersQueue.size()` to be 0 here, right (maybe apart from small window)?